cibuildwheel/__main__.py

```diff
@@ -155,41 +165,43 @@ def main():
     )
 
     if platform == 'linux':
-        manylinux_x86_64_image = os.environ.get('CIBW_MANYLINUX_X86_64_IMAGE', 'manylinux2010')
-        manylinux_i686_image = os.environ.get('CIBW_MANYLINUX_I686_IMAGE', 'manylinux2010')
-        manylinux_pypy_x86_64_image = os.environ.get('CIBW_MANYLINUX_PYPY_X86_64_IMAGE', 'manylinux2010')
-        manylinux_aarch64_image = os.environ.get('CIBW_MANYLINUX_AARCH64_IMAGE', 'manylinux2014')
-        manylinux_ppc64le_image = os.environ.get('CIBW_MANYLINUX_PPC64LE_IMAGE', 'manylinux2014')
-        manylinux_s390x_image = os.environ.get('CIBW_MANYLINUX_S390X_IMAGE', 'manylinux2014')
-
-        default_manylinux_images_x86_64 = {'manylinux1': 'quay.io/pypa/manylinux1_x86_64',
-                                           'manylinux2010': 'quay.io/pypa/manylinux2010_x86_64',
-                                           'manylinux2014': 'quay.io/pypa/manylinux2014_x86_64'}
-        default_manylinux_images_i686 = {'manylinux1': 'quay.io/pypa/manylinux1_i686',
-                                         'manylinux2010': 'quay.io/pypa/manylinux2010_i686',
-                                         'manylinux2014': 'quay.io/pypa/manylinux2014_i686'}
-        default_manylinux_images_pypy_x86_64 = {'manylinux2010': 'pypywheels/manylinux2010-pypy_x86_64'}
-        default_manylinux_images_aarch64 = {'manylinux2014': 'quay.io/pypa/manylinux2014_aarch64'}
-        default_manylinux_images_ppc64le = {'manylinux2014': 'quay.io/pypa/manylinux2014_ppc64le'}
-        default_manylinux_images_s390x = {'manylinux2014': 'quay.io/pypa/manylinux2014_s390x'}
+        pinned_docker_images_file = os.path.join(
+            os.path.dirname(__file__), 'resources', 'pinned_docker_images.cfg'
+        )
+        all_pinned_docker_images = ConfigParser()
+        all_pinned_docker_images.read(pinned_docker_images_file)
+        # all_pinned_docker_images looks like a dict of dicts, e.g.
+        # { 'x86_64': {'manylinux1': '...', 'manylinux2010': '...', 'manylinux2014': '...'},
+        #   'i686': {'manylinux1': '...', 'manylinux2010': '...', 'manylinux2014': '...'},
+        #   'pypy_x86_64': {'manylinux2010': '...' }
+        #   ... }
+
+        manylinux_images = {}
+
+        for build_platform in ['x86_64', 'i686', 'pypy_x86_64', 'aarch64', 'ppc64le', 's390x']:
+            pinned_images = all_pinned_docker_images[build_platform]
+
+            config_name = 'CIBW_MANYLINUX_{}_IMAGE'.format(build_platform.upper())
+            config_value = os.environ.get(config_name)
+
+            if config_value is None:
+                # default to manylinux2010 if it's available, otherwise manylinux2014
+                image = pinned_images.get('manylinux2010') or pinned_images.get('manylinux2014')
+            elif config_value in pinned_images:
+                image = pinned_images[config_value]
+            else:
+                image = config_value
+
+            manylinux_images[build_platform] = image
```

Comment:
I expect that `CIBW_DEPENDENCY_VERSIONS` affect also this. otherwise I expect to have list of all docker images, not only manylinux2010 x86 and 64

Reply:
how would you expect `CIBW_DEPENDENCY_VERSIONS` to affect this? 

> otherwise I expect to have list of all docker images, not only manylinux2010 x86 and 64

i'm not sure I understand. You can only run docker images that match the architecture of the machine you're running on.

Reply:
my bad. I missed list of docker images on begin of documentation section.
 
https://github.com/joerick/cibuildwheel/blob/deterministic-builds/docs/options.md#cibw_manylinux_x86_64_image-cibw_manylinux_i686_image-cibw_manylinux_pypy_x86_64_image-cibw_manylinux_aarch64_image-cibw_manylinux_ppc64le_image-cibw_manylinux_s390x_image--manylinux-image

Reply:
> #cibw_manylinux_x86_64_image-cibw_manylinux_i686_image-cibw_manylinux_pypy_x86_64_image-cibw_manylinux_aarch64_image-cibw_manylinux_ppc64le_image-cibw_manylinux_s390x_image--manylinux-image

Unrelated, but that's a horribly long id. But it's auto-generated, so we can't do anything about that?

Reply:
The docs have a much nicer one, specified by the `{: #manylinux-image}` bit

https://cibuildwheel--256.org.readthedocs.build/en/256/options/#manylinux-image

Reply:
Yes, but it works only in generated docs.